crates/oxc_parser/src/lexer/mod.rs

```diff
@@ -263,14 +263,17 @@ impl<'a> Lexer<'a> {
         self.source.peek_char2()
     }
 
-    /// Peek the next character, and advance the current position if it matches
+    /// Peek the next byte, and advance the current position if it matches
+    /// the given ASCII char.
     #[inline]
-    fn next_eq(&mut self, c: char) -> bool {
-        let matched = self.peek() == Some(c);
-        if matched {
-            self.source.next_char().unwrap();
+    fn next_ascii_char_eq(&mut self, b: u8) -> bool {
+        // TODO: can be replaced by `std::ascii:Char` once stabilized.
+        // https://github.com/rust-lang/rust/issues/110998
+        if !b.is_ascii() {
```

Comment:
Maybe try replacing `if !b.is_ascii() { return false; }` with `assert!(b.is_ascii())`.

You'd think that'd create more verbose assembly, but counter-intuitively, sometimes `assert!` causes compiler to backtrack further and prove that the assertion cannot fail (which it can't if calling code is e.g. `self.next_ascii_char_eq(b'.')`, and therefore removes it entirely, where it won't do so for a normal `if` statement.

Reply:
Also, just as an experiment, you could remove `if !b.is_ascii() { return false; }` and not replace it with anything. This would be completely unsafe, and so could not be the final version, but it'd be interesting to see if that affected benchmarks.

Reply:
I've added both alternatives as separate commits, so you can compare the benchmark results in CodSpeed.
There aren't huge differences among all options, although some lexer and parser cases do see a 1% improvement.
There is also https://github.com/oxc-project/oxc/pull/4304 stacked on top of this and pushing the conversion further along; it shows similar numbers.